v1/status_message.go

```diff
@@ -0,0 +1,147 @@
+package statusproto
+
+import (
+	"crypto/ecdsa"
+	"github.com/pkg/errors"
+	"log"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/golang/protobuf/proto"
+	"github.com/jinzhu/copier"
+	"github.com/status-im/status-protocol-go/applicationmetadata"
+	"github.com/status-im/status-protocol-go/datasync"
+	"github.com/status-im/status-protocol-go/encryption"
+	whisper "github.com/status-im/whisper/whisperv6"
+)
+
+// StatusMessage is any Status Protocol message.
+type StatusMessage struct {
+	// TransportMessage is the parsed message received from the trasport layer, i.e the input
+	TransportMessage *whisper.Message
+	// ParsedMessage is the parsed message by the application layer, i.e the output
+	ParsedMessage interface{}
+
+	// TransportPayload is the payload as received from the transport layer
+	TransportPayload []byte
+	// DecryptedPayload is the payload after having been processed by the encryption layer
+	DecryptedPayload []byte
+
+	// ID is the canonical ID of the message
+	ID []byte
+	// Hash is the transport layer hash
+	Hash []byte
+
+	// TransportLayerSigPubKey contains the public key provided by the transport layer
+	TransportLayerSigPubKey *ecdsa.PublicKey
+	// ApplicationMetadataLayerPubKey contains the public key provided by the application metadata layer
+	ApplicationMetadataLayerSigPubKey *ecdsa.PublicKey
+}
+
+// SigPubKey returns the most important signature, from the application layer to transport
+func (s *StatusMessage) SigPubKey() *ecdsa.PublicKey {
+	if s.ApplicationMetadataLayerSigPubKey != nil {
+		return s.ApplicationMetadataLayerSigPubKey
+	}
+
+	return s.TransportLayerSigPubKey
+}
+
+func (s *StatusMessage) Clone() (*StatusMessage, error) {
+	copy := &StatusMessage{}
+
+	err := copier.Copy(&copy, s)
```

Comment:
Have you tried to just copy the struct? Not sure if this is needed. Pretty much all fields are read only and not modified after initialization.

Reply:
what do you mean exactly by just copying the struct? i.e "manually", or there some othere method?

Reply:
I see what you mean, not sure about it, it's a bit dangerous as the client probably does not expect that (status-go for example would have to know that it makes a shallow copy, as currently it overrides `Payload` of a whisper message, which is a pointer, with the `DecryptedPayload`), and it's a bit difficult to reason about it (those messages that are demuxed are shallow copies), I would probably delay the decision on this one, until the interfaces are a bit more finalized and err on the side of caution (performance wise should not be a problem as I think we rarely batch datasync messages), what do you think?

Reply:
Yeah, we need to be very careful in whatever approach we use so as to not make it easy to screw up later by modifying a shared buffer or something like that unintentionally.

Reply:
We can keep deep copy but generally if a value is received and it has slices, they should be copied by the caller if it wants to modify the underlying array. If we don't have this assumption, we would end up in returning deep copies of slices pretty often.